pilot/pkg/model/service.go

```diff
@@ -45,23 +45,50 @@ type Service struct {
 	// Hostname of the service, e.g. "catalog.mystore.com"
 	Hostname string `json:"hostname"`
 
-	// Address specifies the service IPv4 address of the load balancer
+	// Address specifies the virtual IP address associated with the service.
 	Address string `json:"address,omitempty"`
 
 	// Ports is the set of network ports where the service is listening for
 	// connections
 	Ports PortList `json:"ports,omitempty"`
 
-	// ExternalName is only set for external services and holds the external
-	// service DNS name.  External services are name-based solution to represent
-	// external service instances as a service inside the cluster.
-	ExternalName string `json:"external"`
-
 	// ServiceAccounts specifies the service accounts that run the service.
 	ServiceAccounts []string `json:"serviceaccounts,omitempty"`
 
-	// LoadBalancingDisabled indicates that no load balancing should be done for this service.
-	LoadBalancingDisabled bool `json:"-"`
+	// MeshExternal (if true) indicates that the service is external to the mesh.
+	// These services are defined using Istio's ExternalService spec.
+	MeshExternal bool
+
+	// Virtual (if true) indicates that the service is not discoverable or directly
+	// routable using its hostname. These are typically used to indicate subsets of
+	// a given service (e.g., v1.reviews.bookinfo.com, where v1 is the name of a subset that
+	// indicates the set of instances associated with version v1 of the reviews service.
+	VirtualService bool
```

Comment:
As it stands now, and if I understand correctly, the Service objects and ServiceInstance objects are created by the platform registries as part of the flow of satifying xDS requests. They are all ephemeral (removed after xDS requests). with this proposal, are they going to be part of a cache on top of the platform registries and routing rules? How are the virtual services getting created? Are there going to be some kind of linkage between the virtual services and the destination rules that are defining the subsets? What are the workflows going to be with virtual services involved (for example, creation/deletion of virtual service, destination rule change, how to classify a ServiceInstance into a virtual service, etc)?

Sorry for too many questions. I suggest that we should have an overall design document. It seems to me that everybody agrees to have a cache on top of the platform registries and routing rules, but what's being debated is how to organize the cache so that existing implementation can largely remain unchanged for stability, in the same time new features can be supported with scale. I don't feel like there is a conflict between the two goals since they could be managed with both clear code separation and backward-compatible interfaces.

Reply:
Interesting, @shalako and I were just talking about the opposite: having a virtual service to represent a user-defined name that isn't inherently backed by any particular service, but might (via route-rules) be routable.

Reply:
@baodongli these are great questions. This rfc is not necessarily related to optimization. You could easily compute all the virtual services on the fly without persisting them, at the cost of increased computation. The benefit here is that routing code will no longer deal with labels but just deal with hosts. 

A subsequent step could remove the computation by using a set of hash tables (may be move the ones from mesh.go).

@rosenhouse current route rules already capture this notion (equivalent to cf routes).

Reply:
Yes but we want VIPs for those virtual services, in order to support non-HTTP services used infra-mesh.  How can our platform adapter inform Istio about the mapping of VIP to Istio Service, other than by setting the Address field on the Service?    Would we write a route rule with host match to the VIP?

Reply:
Yes but we want VIPs for those virtual services, in order to support non-HTTP services used infra-mesh.  How can our platform adapter inform Istio about the mapping of VIP to Istio Service, other than by setting the Address field on the Service?    Would we write a route rule with host match to the VIP?